probe-rs/src/probe/blackmagic/mod.rs

```diff
@@ -0,0 +1,1523 @@
+        apsel: u8,
+        addr: u8,
+        value: u32,
+    },
+    MemReadV1 {
+        index: u8,
+        apsel: u8,
+        csw: u32,
+        offset: u32,
+        data: &'a mut [u8],
+    },
+    MemWriteV1 {
+        index: u8,
+        apsel: u8,
+        csw: u32,
+        align: u8,
+        offset: u32,
+        data: &'a [u8],
+    },
+    RawAccessV3 {
+        index: u8,
+        rnw: u8,
+        addr: u8,
+        value: u32,
+    },
+    ReadDpV3 {
+        index: u8,
+        addr: u8,
+    },
+    ReadApV3 {
+        index: u8,
+        apsel: u8,
+        addr: u8,
+    },
+    WriteApV3 {
+        index: u8,
+        apsel: u8,
+        addr: u8,
+        value: u32,
+    },
+    MemReadV3 {
+        index: u8,
+        apsel: u8,
+        csw: u32,
+        offset: u32,
+        data: &'a mut [u8],
+    },
+    MemWriteV3 {
+        index: u8,
+        apsel: u8,
+        csw: u32,
+        align: u8,
+        offset: u32,
+        data: &'a [u8],
+    },
+    MemReadV4 {
+        index: u8,
+        apsel: u8,
+        csw: u32,
+        offset: u64,
+        data: &'a mut [u8],
+    },
+    MemWriteV4 {
+        index: u8,
+        apsel: u8,
+        csw: u32,
+        align: u8,
+        offset: u64,
+        data: &'a [u8],
+    },
+    JtagNext {
+        tms: bool,
+        tdi: bool,
+    },
+    JtagTms {
+        bits: u32,
+        length: usize,
+    },
+    JtagTdi {
+        bits: u32,
+        length: usize,
+        tms: bool,
+    },
+    JtagInit,
+    JtagReset,
+    JtagAddDevice {
+        index: u8,
+        dr_prescan: u8,
+        dr_postscan: u8,
+        ir_len: u8,
+        ir_prescan: u8,
+        ir_postscan: u8,
+        current_ir: u32,
+    },
+    SwdInit,
+    SwdIn {
+        length: usize,
+    },
+    SwdInParity {
+        length: usize,
+    },
+    SwdOut {
+        value: u32,
+        length: usize,
+    },
+    SwdOutParity {
+        value: u32,
+        length: usize,
+    },
+}
+
+impl<'a> RemoteCommand<'a> {
+    /// Return the buffer from the payload of the specified value where the
+    /// response should be written.
+    fn response_buffer(&mut self) -> Option<&mut [u8]> {
+        match self {
+            RemoteCommand::Handshake(data) => Some(data),
+            RemoteCommand::MemReadV0P { data, .. } => Some(data),
+            RemoteCommand::MemReadV1 { data, .. } => Some(data),
+            RemoteCommand::MemReadV3 { data, .. } => Some(data),
+            RemoteCommand::MemReadV4 { data, .. } => Some(data),
+            _ => None,
+        }
+    }
+
+    /// Return `true` if the resulting buffer should have hex decoded.
+    fn decode_hex(&self) -> bool {
+        matches!(
+            self,
+            RemoteCommand::MemReadV0P { .. }
+                | RemoteCommand::MemReadV1 { .. }
+                | RemoteCommand::MemReadV3 { .. }
+                | RemoteCommand::MemReadV4 { .. }
+        )
+    }
+}
+
+impl<'a> core::fmt::Display for RemoteCommand<'a> {
```

Comment:
Display is meant for human consumption and this looks more like the actual command encoding? I think it's fine to implement `ToString` instead.

Reply:
Coming from the embedded world, I still find it odd to be allocating strings all the time. But I agree that it's logically better to implement `ToString` rather than `Display`, so I'll go ahead and do that.